src/main/java/com/couchbase/lite/Database.java

```diff
@@ -1210,6 +1211,111 @@ public synchronized boolean open() {
+                    database.close();
+                    return false;
+                }
+
+                // 2. iterate revs table, and populate attachment table
+                String sql = "SELECT sequence, json FROM revs WHERE no_attachments=0";
+                Cursor cursor2 = null;
+                try {
+                    cursor2 = database.rawQuery(sql, null);
+                    while (cursor2.moveToNext()) {
+                        if (!cursor2.isNull(1)) {
+                            long sequence = cursor2.getLong(0);
+                            byte[] json = cursor2.getBlob(1);
+                            try {
+                                Map<String, Object> docProperties = Manager.getObjectMapper().readValue(json, Map.class);
+                                Log.e(Log.TAG_DATABASE, docProperties.toString());
+                                Map<String, Object> attachments = (Map<String, Object>) docProperties.get("_attachments");
+                                Iterator<String> itr = attachments.keySet().iterator();
+                                while (itr.hasNext()) {
+                                    String name = itr.next();
+                                    Map<String, Object> attachment = (Map<String, Object>) attachments.get(name);
+                                    String contentType = (String) attachment.get("content_type");
+                                    int revPos = (Integer) attachment.get("revpos");
+                                    int length = (Integer) attachment.get("length");
+                                    String digest = (String) attachment.get("digest");
+                                    BlobKey key = new BlobKey(digest);
+                                    try {
+                                        insertAttachmentForSequenceWithNameAndType(sequence, name, contentType, revPos, key, length);
+                                    } catch (CouchbaseLiteException e) {
+                                        Log.e(Log.TAG_DATABASE, "Attachment information inserstion error: " + name + "=" + attachment.toString(), e);
+                                    }
+                                }
+                            } catch (Exception e) {
+                                Log.e(Log.TAG_DATABASE, "JSON parsing error: " + new String(json), e);
+                            }
+                        }
+                    }
+                } catch (SQLException e) {
+                    Log.e(Database.TAG, "Failed to check if attachments table exists", e);
+                    if (cursor2 != null) {
+                        cursor2.close();
+                    }
+                    database.close();
+                    return false;
+                } finally {
+                    if (cursor2 != null) {
+                        cursor2.close();
+                    }
+                }
+            }
+            dbVersion = 20;
```

Comment:
I think if there is an error while migrating, we should rollback everything (including the created attachments table and schema version).